cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/DefaultApplicationUpdateContext.java

```
@@ -0,0 +1,164 @@
+  // TODO: Pass ArtifactSortOrder as argument for better flexibility.
+  @Override
+  public List<ArtifactId> getPluginArtifacts(String pluginType, String pluginName, ArtifactScope pluginScope,
+                                             @Nullable ArtifactVersionRange pluginRange, int limit) {
+    List<ArtifactId> pluginArtifacts = new ArrayList<>();
+    NamespaceId pluginArtifactNamespace = ArtifactScope.SYSTEM.equals(pluginScope) ? NamespaceId.SYSTEM : namespaceId;
+
+    Predicate<io.cdap.cdap.proto.id.ArtifactId> predicate = new Predicate<io.cdap.cdap.proto.id.ArtifactId>() {
+      @Override
+      public boolean apply(io.cdap.cdap.proto.id.ArtifactId input) {
+        // should check if the artifact is from SYSTEM namespace, if not, check if it is from the scoped namespace.
+        // should check if plugin is in given range if provided.
+        return (((pluginScope == null && NamespaceId.SYSTEM.equals(input.getParent()))
+               || pluginArtifactNamespace.equals(input.getParent())) &&
+               (pluginRange == null || pluginRange.versionIsInRange(new ArtifactVersion(input.getVersion()))));
+      }
+    };
+
+    try {
+      Map<ArtifactDescriptor, PluginClass> plugins =
+          artifactRepository.getPlugins(pluginArtifactNamespace,
+                                        Artifact.from(Namespace.fromEntityId(namespaceId), applicationArtifactId),
+                                        pluginType, pluginName, predicate, limit, ArtifactSortOrder.ASC);
+      for (Map.Entry<ArtifactDescriptor, PluginClass> pluginsEntry : plugins.entrySet()) {
+        ArtifactId plugin = pluginsEntry.getKey().getArtifactId();
+        // Only consider non-SNAPSHOT plugins for upgrade.
+        // TODO: Consider making this check optional. Helpful for integration tests.
+        if (!plugin.getVersion().isSnapshot()) {
+          pluginArtifacts.add(plugin);
+        }
+      }
+    } catch (Exception e) {
```

exceptions getting plugins should not be swallowed and ignored. They need to be propagated up and result in update failure.

you probably need to update the interface and potentially add new exception classes to the cdap-api.

---

I agree. Right now I log and throw exception above.. it will result in upgrade failure for sure now. But for now, I am catching and throwing generic exception. I will follow up with new upgrade related exception class in cdap-api in a follow up PR after tracing down all potential exceptions (I tried to do it right now but thought it might be better and cleaner to do it in follow up) if that works for now. Let me know your thoughts.